### DISCUSSION TITLE
Топ треков от исполнителей добавленных в избранное

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#28] [Помощь с алгоритмом] [answered]
Прошу помощи в написании куска скрипта. Пытался сделать сам, используя примеры здесь и на 4PDA, но так и не осилил.

Нужно из списка треков (я использую список треков, добавленных день назад в избранное), получить список состоящий из топов треков этих исполнителей. От каждого исполнителя по одному треку. Треки должны отсутствовать в истории прослушивания и в любимых треках (ну с этим все понятно). Так же треки не должны быть ниже третьего места в топе исполнителя.
Основной затык - как из списка треков получить список исполнителей. 

Кусок будет использоваться в ежедневном скрипте для более детального ознакомления со свежеоткрытыми исполнителями.

**Comment by Chimildic** [comment, marked as answer]
> как из списка треков получить список исполнителей

Добавить следующий `map` к массиву треков. Вернутся исполнители.
```js
let artists = tracks.map(t => t.artists[0]);
// Можно цепочкой
// let artists = Source.get*().map(t => t.artists[0]);
```
> получить список состоящий из топов треков этих исполнителей
```js
let topTracks = Source.getArtistsTopTracks(artists, false);
```
> Так же треки не должны быть ниже третьего места в топе
```js
topTracks.forEach(tracks => tracks.splice(3, 7)); // удалить треки с 4 по 10
topTracks = topTracks.flat(1); // треки в один массив, подробнее в описании getArtistsTopTracks про flat
```

**Comment by Bedmit** [comment]
Отлично! То что и хотел. Возможно кому-то будет интересен результат
```
function Test() {
  let Finding = Source.getSavedTracks();
  Filter.rangeDateRel(Finding, 1, 1);
  let Artists = Finding.map(t => t.artists[0]);
  let FindCheck = Source.getArtistsTopTracks(Artists, false);
  FindCheck.forEach(tracks => tracks.splice(3, 7));
  FindCheck = FindCheck.flat(1);
  let History = Cache.read('AllHistory');
  let Past = Source.getSavedTracks();
  Filter.removeTracks(FindCheck, History);
  Filter.removeTracks(FindCheck, Past);
  Filter.dedupArtists(FindCheck);

  Playlist.saveAsNew({
  name: 'Test',
  tracks: FindCheck,
});
}
```
